aten/src/ATen/native/Distance.cpp

```diff
@@ -83,10 +83,13 @@ Tensor _pdist_backward(const Tensor& grad, const Tensor& self, const double p, c
 }
 
 Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double eps) {
+  // Follow scipy impl to improve numerical precision
+  // Use x / sqrt(x * x) instead of x / (sqrt(x) * sqrt(x))
   Tensor w12 = at::sum(x1 * x2, dim);
-  Tensor w1 = at::norm(x1, 2, dim);
-  Tensor w2 = at::norm(x2, 2, dim);
-  return w12.div_((w1 * w2).clamp_min_(eps));
+  Tensor w1 = at::sum(x1 * x1, dim);
+  Tensor w2 = at::sum(x2 * x2, dim);
+  Tensor n12 = (w1 * w2).sqrt_().clamp_min(eps);
```

Comment:
arguably we should use rsqrt here for better precision :)

Reply:
This makes the precision worse.

On CPU, rsqrt(x) is implemented as `1/sqrt(x)` so now three operations with rounding instead of two. (There's no std::rsqrt in C++. The x86 rsqrt instructions are low-precision).

With CUDA it's a little different, but sqrt(x) is generally more precise on modern GPUs.

Reply:
Thanks @colesbury ! I will send a patch to fix this.